### PR TITLE
nixos/peertube: add nginx configuration

### PR DESCRIPTION
**Pull request by Izorkin**
###### Description of changes
Add nginx configuration.

cc @marsam @mohe2015 @immae @SuperSandro2000

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/peertube.nix [hunk 1]
```diff
@@ -441,6 +449,293 @@ in {
...
+            proxy_set_header X-Real-IP                  $remote_addr;
+            proxy_set_header Upgrade                    $http_upgrade;
+            proxy_set_header Connection                 'upgrade';
+
+            proxy_http_version                          1.1;
+          '';
+        };
+
+        # Bypass PeerTube for performance reasons.
+        locations."~ ^/client/(assets/images/(icons/icon-36x36\.png|icons/icon-48x48\.png|icons/icon-72x72\.png|icons/icon-96x96\.png|icons/icon-144x144\.png|icons/icon-192x192\.png|icons/icon-512x512\.png|logo\.svg|favicon\.png|default-playlist\.jpg|default-avatar-account\.png|default-avatar-account-48x48\.png|default-avatar-video-channel\.png|default-avatar-video-channel-48x48\.png))$" = {
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/peertube.nix:544]: Can we break this line somewhere?

**Review by SuperSandro2000 — COMMENTED** [review]
no idea if the settings are correct though

File: nixos/modules/services/web-apps/peertube.nix (hunk 1 above)
Comment (Izorkin) [reply nixos/modules/services/web-apps/peertube.nix:544]: Will not work. Only if split into parts.

**Comment by Izorkin** [comment]
Added small fixes.
cc @SuperSandro2000

File: nixos/modules/services/web-apps/peertube.nix [hunk 2]
```diff
@@ -441,6 +452,293 @@ in {
...
+          priority = 1230;
+
+          extraConfig = ''
+            proxy_set_header X-Forwarded-For            $proxy_add_x_forwarded_for;
+            proxy_set_header Host                       $host;
+            proxy_set_header X-Real-IP                  $remote_addr;
+            proxy_set_header Upgrade                    $http_upgrade;
+            proxy_set_header Connection                 'upgrade';
+
+            proxy_http_version                          1.1;
```
[the remark this answers is not in the document]
Comment (Izorkin) [reply nixos/modules/services/web-apps/peertube.nix:548]: Value of `proxy_set_header Connection` parameter is different there - `$connection_upgrade;`

File: nixos/modules/services/web-apps/peertube.nix [hunk 3]
```diff
@@ -441,6 +452,293 @@ in {
...
+              add_header Access-Control-Allow-Headers   'Range,DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type';
+              add_header Access-Control-Max-Age         1728000;
+              add_header Content-Type                   'text/plain charset=UTF-8';
+              add_header Content-Length                 0;
+              return                                    204;
+            }
+            if ($request_method = 'GET') {
+              add_header Access-Control-Allow-Origin    '*';
+              add_header Access-Control-Allow-Methods   'GET, OPTIONS';
+              add_header Access-Control-Allow-Headers   'Range,DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type';
```
[the remark this answers is not in the document]
Comment (Izorkin) [reply nixos/modules/services/web-apps/peertube.nix]: Created.

File: nixos/modules/services/web-apps/peertube.nix [hunk 4]
```diff
@@ -441,6 +458,269 @@ in {
...
+            ${nginxCommonHeaders}
+            add_header Cache-Control                    'public, max-age=7200';
+
+            rewrite ^/static/thumbnails/(.*)$           /$1 break;
+          '';
+        };
+
+        locations."^~ /static/redundancy/" = {
+          tryFiles = "$uri @api";
+          root = "${cfg.settings.storage.redundancy}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/peertube.nix:632]:
```suggestion
          root = cfg.settings.storage.redundancy;
```

File: nixos/modules/services/web-apps/peertube.nix [hunk 5]
```diff
@@ -441,6 +458,269 @@ in {
...
+            ${nginxCommonHeaders}
+            add_header Cache-Control                    'public, max-age=7200';
+
+            rewrite ^/lazy-static/previews/(.*)$        /$1 break;
+          '';
+        };
+
+        locations."^~ /static/thumbnails/" = {
+          tryFiles = "$uri @api";
+          root = "${cfg.settings.storage.thumbnails}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/peertube.nix:611]:
```suggestion
          root = cfg.settings.storage.thumbnails;
```

File: nixos/modules/services/web-apps/peertube.nix [hunk 6]
```diff
@@ -441,6 +458,269 @@ in {
...
+            add_header Cache-Control                    'public, max-age=7200';
+
+            rewrite ^/lazy-static/avatars/(.*)$         /$1 break;
+            rewrite ^/lazy-static/banners/(.*)$         /$1 break;
+          '';
+        };
+
+        locations."^~ /lazy-static/previews/" = {
+          tryFiles = "$uri @api";
+          root = "${cfg.settings.storage.previews}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/peertube.nix:590]:
```suggestion
          root = cfg.settings.storage.previews;
```

File: nixos/modules/services/web-apps/peertube.nix [hunk 7]
```diff
@@ -441,6 +458,269 @@ in {
...
+          alias = "${cfg.package}/client/dist/$1";
+          priority = 1320;
+          extraConfig = ''
+            add_header Cache-Control                    'public, max-age=604800, immutable';
+          '';
+        };
+
+        locations."~ ^/lazy-static/(avatars|banners)/" = {
+          tryFiles = "$uri @api";
+          root = "${cfg.settings.storage.avatars}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/peertube.nix:568]:
```suggestion
          root = cfg.settings.storage.avatars;
```

File: nixos/modules/services/web-apps/peertube.nix [hunk 8]
```diff
@@ -441,6 +458,269 @@ in {
...
+          tryFiles = "/dev/null @api_websocket";
+          priority = 1220;
+
+          extraConfig = ''
+            proxy_read_timeout                          15m;
+          '';
+        };
+
+        locations."@api_websocket" = {
+          proxyPass = "http://127.0.0.1:${toString(cfg.listenHttp)}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/peertube.nix:538]:
```suggestion
          proxyPass = "http://127.0.0.1:${toString cfg.listenHttp}";
```

File: nixos/modules/services/web-apps/peertube.nix [hunk 9]
```diff
@@ -441,6 +458,269 @@ in {
...
+          priority = 1140;
+
+          extraConfig = ''
+            client_max_body_size                        6M;
+            add_header X-File-Maximum-Size              4M always;
+          '';
+        };
+
+        locations."@api" = {
+          proxyPass = "http://127.0.0.1:${toString(cfg.listenHttp)}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/peertube.nix:504]:
```suggestion
          proxyPass = "http://127.0.0.1:${toString cfg.listenHttp}";
```

File: nixos/modules/services/web-apps/peertube.nix [hunk 10]
```diff
@@ -441,6 +458,269 @@ in {
...
+
+          extraConfig = ''
+            client_max_body_size                        0;
+            proxy_request_buffering                     off;
+          '';
+        };
+
+        locations."~ ^/api/v1/videos/(upload|([^/]+/studio/edit))$" = {
+          tryFiles = "/dev/null @api";
+          root = "${cfg.settings.storage.tmp}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/peertube.nix:484]:
```suggestion
          root = cfg.settings.storage.tmp;
```

**Review by SuperSandro2000 — COMMENTED** [review]
LGTM with those last formatting changes.

I assume you tested this in a deployment? If no one else has anything to add I would merge this after this.

**Comment by Izorkin** [comment]
@SuperSandro2000 fixed!

**Pull request merged**